src/util/sighash.rs

```diff
@@ -0,0 +1,1241 @@
+    /// [std::io::Write] trait.
+    pub fn segwit_encode_signing_data_to<Write: io::Write>(
+        &mut self,
+        mut writer: Write,
+        input_index: usize,
+        prevouts: &Prevouts,
+        sighash_type: SigHashType,
+    ) -> Result<(), Error> {
+        prevouts.check_all(&self.tx)?;
+        self.check_index(input_index)?;
+
+        let zero_hash = sha256d::Hash::default();
+
+        let (sighash, anyone_can_pay) = sighash_type.split_anyonecanpay_flag();
+
+        self.tx.version.consensus_encode(&mut writer)?;
+
+        if !anyone_can_pay {
+            self.segwit_cache().prevouts.consensus_encode(&mut writer)?;
+        } else {
+            zero_hash.consensus_encode(&mut writer)?;
+        }
+
+        if !anyone_can_pay && sighash != SigHashType::Single && sighash != SigHashType::None {
+            self.segwit_cache()
+                .sequences
+                .consensus_encode(&mut writer)?;
+        } else {
+            zero_hash.consensus_encode(&mut writer)?;
+        }
+
+        {
+            let txin = &self.tx.input[input_index];
+
+            txin.previous_output.consensus_encode(&mut writer)?;
+            prevouts
+                .get(input_index)?
+                .script_pubkey
+                .consensus_encode(&mut writer)?;
+            prevouts
+                .get(input_index)?
+                .value
+                .consensus_encode(&mut writer)?;
+            txin.sequence.consensus_encode(&mut writer)?;
+        }
+
+        if sighash != SigHashType::Single && sighash != SigHashType::None {
+            self.segwit_cache().outputs.consensus_encode(&mut writer)?;
+        } else if sighash == SigHashType::Single && input_index < self.tx.output.len() {
+            let mut single_enc = SigHash::engine();
+            self.tx.output[input_index].consensus_encode(&mut single_enc)?;
+            SigHash::from_engine(single_enc).consensus_encode(&mut writer)?;
+        } else {
+            zero_hash.consensus_encode(&mut writer)?;
+        }
+
+        self.tx.lock_time.consensus_encode(&mut writer)?;
+        sighash_type.as_u32().consensus_encode(&mut writer)?;
+        Ok(())
+    }
+
+    /// Compute the BIP143 sighash for any flag type.
+    pub fn segwit_signature_hash(
+        &mut self,
+        input_index: usize,
+        prevouts: &Prevouts,
+        sighash_type: SigHashType,
+    ) -> Result<SigHash, Error> {
+        let mut enc = SigHash::engine();
+        self.segwit_encode_signing_data_to(&mut enc, input_index, prevouts, sighash_type)?;
+        Ok(SigHash::from_engine(enc))
+    }
+
+    /// Encode the legacy signing data for any flag type into a given object implementing a
+    /// [std::io::Write] trait. Internally calls [Transaction::encode_signing_data_to]
+    pub fn legacy_encode_signing_data_to<Write: io::Write, U: Into<u32>>(
+        &self,
+        mut writer: Write,
+        input_index: usize,
+        prevouts: &Prevouts,
+        sighash_type: U,
```

Comment:
Wanted to change to SigHashType but it's incompatible with `let sighash_arr = endian::u32_to_array_le(sighash_type);` inside `Transaction::encode_signing_data_to`